File: src/linkerz_crawler_core/src/main/java/org/linkerz/crawler/core/parser/DefaultParser.java
/*
 * Copyright (C) 2012 - 2013 LinkerZ
 */

package org.linkerz.crawler.core.parser;

import edu.uci.ics.crawler4j.parser.ExtractedUrlAnchorPair;
import edu.uci.ics.crawler4j.parser.HtmlContentHandler;
import edu.uci.ics.crawler4j.url.URLCanonicalizer;
import org.apache.tika.metadata.Metadata;
import org.apache.tika.parser.ParseContext;
import org.apache.tika.parser.html.HtmlParser;
import org.linkerz.crawler.core.downloader.result.DefaultDownloadResult;
import org.linkerz.crawler.core.model.WebLink;
import org.linkerz.crawler.core.parser.result.DefaultParserResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * The Class DefaultParser.
 *
 * @author Nguyen Duc Dung
 * @since 7/2/12, 2:15 AM
 */
public class DefaultParser extends AbstractParser<DefaultDownloadResult, DefaultParserResult> {

    private final static Logger logger = LoggerFactory.getLogger(DefaultParser.class);

    private HtmlParser htmlParser = new HtmlParser();
    private ParseContext parseContext = new ParseContext();

    @Override
    public DefaultParserResult parse(DefaultDownloadResult downloadResult) {
        List<WebLink> webLinks = new ArrayList<WebLink>();
        if (downloadResult.getWebPage() != null) {
            Metadata metadata = new Metadata();
            HtmlContentHandler contentHandler = new HtmlContentHandler();

            InputStream inputStream = null;
            try {
                inputStream = new ByteArrayInputStream(downloadResult.getWebPage().getHtml().getBytes());
                htmlParser.parse(inputStream, contentHandler, metadata, parseContext);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                try {
                    if (inputStream != null) {
                        inputStream.close();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            downloadResult.getWebPage().setTitle(metadata.get(Metadata.TITLE));

            String baseURL = contentHandler.getBaseUrl();
            String contextURL = downloadResult.getWebPage().getWebLink().getUrl();
            if (baseURL != null) {
                contextURL = baseURL;
            }

            for (ExtractedUrlAnchorPair urlAnchorPair : contentHandler.getOutgoingUrls()) {
                String href = urlAnchorPair.getHref();
                href = href.trim();
                if (href.length() == 0) {
                    continue;
                }
                String hrefWithoutProtocol = href.toLowerCase();
                if (href.startsWith("http://")) {
                    hrefWithoutProtocol = href.substring(7);
                }
                if (!hrefWithoutProtocol.contains("javascript:") && !hrefWithoutProtocol.contains("@")) {
                    String url = URLCanonicalizer.getCanonicalURL(href, contextURL);
                    if (url != null) {
                        webLinks.add(new WebLink(url));
                    }
                }
            }
        }

        return new DefaultParserResult(webLinks);
    }
}

File: src/linkerz_crawler_core/src/main/scala/org/linkerz/crawler/core/parser/DefaultParser.scala
/*
 * Copyright (C) 2012 - 2013 LinkerZ
 */

package org.linkerz.crawler.core.parser

import org.apache.tika.parser.html.HtmlParser
import org.apache.tika.parser.ParseContext
import edu.uci.ics.crawler4j.parser.HtmlContentHandler
import collection.mutable.ListBuffer
import org.linkerz.crawler.core.model.WebUrl
import org.apache.tika.metadata.Metadata
import java.io.ByteArrayInputStream
import edu.uci.ics.crawler4j.url.URLCanonicalizer
import collection.JavaConversions._
import org.linkerz.crawler.core.job.CrawlJob
import org.apache.commons.validator.routines.UrlValidator

/**
 * The Class DefaultParser.
 *
 * @author Nguyen Duc Dung
 * @since 8/12/12, 11:04 PM
 *
 */

class DefaultParser extends Parser {

  val htmlParser = new HtmlParser
  val parseContext = new ParseContext
  val htmlHandler = new HtmlContentHandler

  def parse(crawlJob: CrawlJob) {
    val webUrl = crawlJob.result.get.webUrl
    val webPage = crawlJob.result.get

    info("Parse: " + webUrl.url)

    var webUrls = new ListBuffer[WebUrl]

    if (webPage.content != null) {
      val metadata = new Metadata
      val inputStream = new ByteArrayInputStream(webPage.content)
      htmlParser.parse(inputStream, htmlHandler, metadata, parseContext)

      webPage.title = metadata.get(Metadata.TITLE)
      webPage.contentEncoding = metadata.get("Content-Encoding")

      //Get web page content
      //      val title = metadata.get(Metadata.TITLE)
      //      val subTitle = metadata.get(Metadata.DESCRIPTION)
      //      val html = htmlHandler.getBodyText

      //      if (DetectorFactory.getLangList.isEmpty) {
      //        DetectorFactory.loadProfile(new File(Resources.getResource("profiles").toURI))
      //      }
      //
      //      val detector = DetectorFactory.create
      //      detector.setMaxTextLength(1000)
      //      detector.append(html)
      //
      //      try {
      //        val language = detector.detect
      //        webPage.language = language
      //      } catch {
      //        case ex: Exception => error(ex.getMessage, ex)
      //      }

      //Extract links in side a website
      val baseURL = htmlHandler.getBaseUrl
      var contextURL = webUrl.url
      if (baseURL != null) {
        contextURL = baseURL
      }

      htmlHandler.getOutgoingUrls.foreach(urlAnchorPair => {
        var href = urlAnchorPair.getHref
        href = href.trim()
        if (href.length() != 0) {
          var hrefWithoutProtocol = href.toLowerCase
          if (href.startsWith("http://")) {
            hrefWithoutProtocol = href.substring(7)
          }
          if (!hrefWithoutProtocol.contains("javascript:")
            && !hrefWithoutProtocol.contains("@")
            && !hrefWithoutProtocol.contains("mailto:")) {
            val url = URLCanonicalizer.getCanonicalURL(href, contextURL)
            val urlValidator = new UrlValidator(Array("http","https"))
            if (url != null && urlValidator.isValid(url)) {
              webUrls += new WebUrl(url)
            }
          }
        }
      })
    }

    webPage.webUrls = webUrls.toList
  }

}
